### PR TITLE
port `LFValidator.xtend` to java

### PR DESCRIPTION
**Pull request by housengw**
see #838

### CODE REVIEW

**Comment by lhstrh** [comment]
The validator is a very important class, and it will be difficult to assure the ported code is functionally correct. I would suggest that before merging this in, we improve code coverage of `LFValidator` with unit tests.

**Comment by lhstrh** [comment]
@housengw could you incorporate the following changes in your port please?
https://github.com/lf-lang/lingua-franca/pull/804/files
In your commit message, please reference PR #804 so that is shows up in the conversation of that PR.
Once this has been done, I will close #804.

**Comment by lhstrh** [comment]
The dummy Xtend file shouldn't be needed if you set `generateXtendStub = true` to `generateXtendStub = false` in `org.lflang/src/org/lflang/GenerateLinguaFranca.mwe2`

**Review by cmnrd — APPROVED** [review]
This looks great! Due to the size of the Validator, it's not possible to check every line, but I made a quick read through. One issue that I noticed, is that the `checkWidthSpec` method disappeared in the Java version.

Also, when opening the java file in IDEA, there is a ton of warnings reported. Many of which suggest code improvements. This is not to say that your code is bad, but more to say that we need to think about our developer setups and what are the recommended tools for working with our code base should be. Luckily, when we don't rely on xtend anymore, we have more options then eclipse :)

**Comment by cmnrd** [comment]
I think eventually we will anyway need to do another pass over the core classes. There are lots of opportunities for designing the validator in a cleaner and more modular way. In such a go we could also modernize the Java code or, if we don't need to rely on Eclipse for development anymore (which should be the case once we have thrown out all xtend), we could also port the class entirely to Kotlin. With IntelliJ's tools for auto-conversion and code linting, this should be a walk in the park compared to the transition from xtend to java ;)

File: org.lflang/src/org/lflang/validation/LFValidator.xtend
Comment (housengw) [reply org.lflang/src/org/lflang/validation/LFValidator.xtend:342]: thanks for reviewing the code. That's a good catch. I have readded the function in the Java version.

**Comment by lhstrh** [comment]
Adding a link to [my comment](https://github.com/lf-lang/lingua-franca/pull/873/files#r794972918) in #873.

**Comment by lhstrh** [comment]
I think we should merge this after tagging `0.1.0-beta`.

**Pull request merged**
